File: Python/1448_Count_Good_Nodes_in_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node, maxval):
            if not node:
                return 0 
            
            if node.val>=maxval:
                count=1
                maxval=node.val
            else:
                count=0

            count+=dfs(node.left,maxval)
            count+=dfs(node.right,maxval)

            return count

        return dfs(root, float('-inf'))

File: Java/1448_Count_Good_Nodes_in_Binary_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        Queue<Pair<TreeNode,Integer>> q=new LinkedList<>();
        int res=0;
        q.add(new Pair<TreeNode,Integer>(root,root.val));
        while(!q.isEmpty()){
            int size=q.size();
            for(int i=0;i<size;i++){
                Pair<TreeNode,Integer> t=q.poll();
                TreeNode node=t.getKey();
                int v=t.getValue();
                if(node!=null){
                    res+= node.val>=v?1:0;
                    int maxv=Math.max(node.val,v);
                    q.add(new Pair<TreeNode,Integer>(node.left,maxv));
                    q.add(new Pair<TreeNode,Integer>(node.right,maxv));
                }    
            }
        }
        return res;    
    }
}
/*  public int goodNodes(TreeNode root) {
        return dfs(root,root.val);
    }
    private int dfs(TreeNode node,int maxValue){
        if(node==null){return 0;}
        int res=node.val>=maxValue?1:0;
        maxValue=Math.max(node.val,maxValue);

        res+=dfs(node.left,maxValue);
        res+=dfs(node.right,maxValue);
        return res;
    }
}*/
